handle redirect

yingjin opened the pull request:
## References
* Fixes #8699 

## Description
This fix helps to preserve handles for contents moved from DSpace to another platform. Right now, all DSpace handles resolved as DSpace URL.

## Instructions for Reviewers

List of changes in this PR:
Adding a new table called handle redirect with following structure. 
    handle_redirect_id  INTEGER PRIMARY KEY DEFAULT nextval('handle_redirect_id_seq'),
    handle              VARCHAR(256) UNIQUE,
    url                     VARCHAR(2048) UNIQUE
* Updated dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java
* Added dspace-api/src/main/java/org/dspace/handleredirect/HandleRedirect.java
               dspace-api/src/main/java/org/dspace/handleredirect/HandleRedirectServiceImpl.java
               dspace-api/src/main/java/org/dspace/handleredirect/dao/HandleRedirectDAO.java
               dspace-api/src/main/java/org/dspace/handleredirect/dao/impl/HandleRedirectDAOImpl.java
               dspace-api/src/main/java/org/dspace/handleredirect/factory/HandleRedirectServiceFactory.java
               dspace-api/src/main/java/org/dspace/handleredirect/factory/HandleRedirectServiceFactoryImpl.java
               dspace-api/src/main/java/org/dspace/handleredirect/service/HandleRedirectService.java
               dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V7.6_2023.03.06__add_table_handleredirect.sql

**Include guidance for how to test or review your PR.** 

You need a handle prefix or setup an independent handle server to do the testing.  Work for Postgresql database only.

Follow DSpace 7 installation instructions for creating dspace database using postgresql.
Compile and ant fresh_install the code. Then run [dspace-install-dir]/bin/dspace database migrate
This will create the handleredirect table.

Start handle server.

Create an administrator account

Login to DSpace UI and add new items. 

For example, the new item handle is 123456789/3.

> psql -d dspace
> Insert into handleredirect (handle_redirect_id, handle, url) values (default, '123456789/3', 'URL of your choice');

Then you can go and test the handle url and see if it redirect to your choice of URL inserted into the handleredirect table.

The fix is to check handleredirect table first, if it is in that table, redirect happens. Otherwise, generate DSpace url.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ x ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [ ] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [ ] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

yingjin committed:
handle redirect

yingjin committed:
update to pass the UnitTest and integration-test

yingjin committed:
Merge branch 'main' into DS-8699

bkeese commented:
@yingjin I was hoping to use this PR to manage some handle redirects in our DSpace 7x instance. I successfully merged the PR and rebuilt DSpace. I ran the database migration and added a test entry to the new handleredirect table.
When I try to go to the handle of that test entry (https://hdl.handle.net/1234/1234), it isn't redirected to the specified url, but instead goes to the dspace item page. I have added debugging in the code where it asks the handleredirectservice for a url, but that code never seems to be run. Any idea if code changes in the last 2 years might have broken this?

yingjin commented:
> @yingjin I was hoping to use this PR to manage some handle redirects in our DSpace 7x instance. I successfully merged the PR and rebuilt DSpace. I ran the database migration and added a test entry to the new handleredirect table. When I try to go to the handle of that test entry (https://hdl.handle.net/1234/1234), it isn't redirected to the specified url, but instead goes to the dspace item page. I have added debugging in the code where it asks the handleredirectservice for a url, but that code never seems to be run. Any idea if code changes in the last 2 years might have broken this?

May I know which version you are using? I have it on DSpace 7.6 and still working well. If it didn't ask for handleredirectservice, likely DSpace took over it before handleredict. Make sure you check redirect table before DSpace handle the url. Here is part of the code in HandlePlugin.java -

            // check if the handle should be redirected before solve
            // it as DSpace handle

            String url = handleRedirectService.resolveToURL(context, handle);
            if (url == null) {
                url = handleService.resolveToURL(context, handle);
                if (url == null) {
                    return null;
                }
            }

Also, make sure the handle server is running. The code can only be executed when the url is redirected from a handle server. If your '1234' is not a real handle prefix, it won't work.

Here is an example: https://hdl.handle.net/1911/81426. The url will redirect an old DSpace item to its new location in Quartex. 

Hope this helps.
Ying

bkeese commented:
Thank you so much for you fast response!  I see the problem. We are also running 7.6, but we are using an external handle server and I didn't update the plugin code that it is using. (https://wiki.lyrasis.org/display/DSDOC7x/Handle.Net+Registry+Support#Handle.NetRegistrySupport-ToinstallaHandleresolveronaseparatemachine) I'll do that and post back here.

yingjin commented:
Great! Looking forward to your update.

On Mon, Apr 28, 2025 at 12:42 PM Brian Keese ***@***.***>
wrote:

> *bkeese* left a comment (DSpace/DSpace#8702)
> <https://github.com/DSpace/DSpace/pull/8702#issuecomment-2836008254>
>
> Thank you so much for you fast response! I see the problem. We are also
> running 7.6, but we are using an external handle server and I didn't update
> the plugin code that it is using. (
> https://wiki.lyrasis.org/display/DSDOC7x/Handle.Net+Registry+Support#Handle.NetRegistrySupport-ToinstallaHandleresolveronaseparatemachine)
> I'll do that and post back here.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DSpace/DSpace/pull/8702#issuecomment-2836008254>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALAW7QALK7M5ZSWWKHNJQ323ZR55AVCNFSM6AAAAAB4A32PRSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZWGAYDQMRVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bkeese commented:
I found it. I had to make a similar change to:
dspace-api/src/main/java/org/dspace/handle/hdlresolver/HdlResolverServiceImpl.java

`import org.dspace.handleredirect.service.HandleRedirectService;

    @Autowired(required = true)
    private HandleRedirectService handleRedirectService;

    @Override
    public String resolveToURL(Context context, HdlResolverDTO hdlResolver) {
        try {
            String url = this.handleRedirectService.resolveToURL(context, hdlResolver.getHandle());
            if (url != null) {
                log.info("GOT REDIRECT URL FOR HANDLE "+hdlResolver.getHandle()+": "+url);
            } else {
                url =  this.handleService.resolveToURL(context, hdlResolver.getHandle());
            }
            return url;
        } catch (SQLException e) {
            log.error("Error while resolving Handle: " + hdlResolver.getHandle(), e);
            throw new RuntimeException("Error while resolving Handle: " + hdlResolver.getHandle(), e);
        }
    }`

yingjin commented:
Awesome!

On Mon, Apr 28, 2025 at 7:36 PM Brian Keese ***@***.***>
wrote:

> *bkeese* left a comment (DSpace/DSpace#8702)
> <https://github.com/DSpace/DSpace/pull/8702#issuecomment-2837124050>
>
> I found it. I had to make a similar change to:
>
> dspace-api/src/main/java/org/dspace/handle/hdlresolver/HdlResolverServiceImpl.java
>
> `import org.dspace.handleredirect.service.HandleRedirectService;
>
> @Autowired(required = true)
> private HandleRedirectService handleRedirectService;
>
> @Override
> public String resolveToURL(Context context, HdlResolverDTO hdlResolver) {
>     try {
>         String url = this.handleRedirectService.resolveToURL(context, hdlResolver.getHandle());
>         if (url != null) {
>             log.info("GOT REDIRECT URL FOR HANDLE "+hdlResolver.getHandle()+": "+url);
>         } else {
>             url =  this.handleService.resolveToURL(context, hdlResolver.getHandle());
>         }
>         return url;
>     } catch (SQLException e) {
>         log.error("Error while resolving Handle: " + hdlResolver.getHandle(), e);
>         throw new RuntimeException("Error while resolving Handle: " + hdlResolver.getHandle(), e);
>     }
> }`
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DSpace/DSpace/pull/8702#issuecomment-2837124050>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALAW7VZSPVRLTY2J3RPXCD233CQRAVCNFSM6AAAAAB4A32PRSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZXGEZDIMBVGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

yingjin commented:
Hi Tom,

I have migrated the handle redirect to DSpace 8 on our local repository,
and it works fine. I'm not sure if it will work on DSpace 9, though.

We deleted it in DSpace after those contents migrated to Quartex, and
recorded those handles with the new Quartex URL in the handleredirect table.

Hope this helps.

Best,
Ying




On Fri, Oct 17, 2025 at 9:14 AM Tom Misilo ***@***.***> wrote:

> *misilot* left a comment (DSpace/DSpace#8702)
> <https://github.com/DSpace/DSpace/pull/8702#issuecomment-3415788715>
>
> @yingjin <https://github.com/yingjin> I am interested in trying this out!
> Do you know if it works with DSpace 8 or 9?
>
> Also, just curious if you just leave the older content in DSpace, or do
> you delete it after migrating, and just keep the redirects in place?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DSpace/DSpace/pull/8702#issuecomment-3415788715>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALAW7Q6JRKCQ4WSCYTB3Q33YD2TJAVCNFSM6AAAAAB4A32PRSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTIMJVG44DQNZRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
